### ISSUE TITLE
Let resources having an Identifier extend a common interface

### ISSUE THREAD

**Issue by taconaut** [FirelyTeam/firely-net-sdk#2205]
I have to access the `Identifier` property of different resources (`MedicationStatement`, `MedicationDispense`, `MedicationRequest`, `Observation`). This property is not on a base class or available on an interface, so I have to handle every case explicitely, whereas I'd like to make it generic.

Could you add an interface `IIdentifiable` (like the already existing `IExtendable`) and make all classes implementing `Identifier` extend it?

Here's the code I've written, which led me to opening this feture request:
```c#
    private static bool IsResourceContainedInRetrievals(DomainResource resource, IEnumerable<MedicamentRetrieval> retrievals)
    {
        if (resource is MedicationStatement medicationStatement)
            return IsMedicationStatementContainedInRetrievals(medicationStatement, retrievals);

        if (resource is MedicationDispense medicationDispense)
            return IsMedicationDispenseContainedInRetrievals(medicationDispense, retrievals);

        if (resource is MedicationRequest medicationRequest)
            return IsMedicationRequestContainedInRetrievals(medicationRequest, retrievals);

        if (resource is Observation observation)
            return IsObservationContainedInRetrievals(observation, retrievals);

        throw new NotSupportedException($"Resource of type='{(resource == null ? "NULL" : resource.GetType().Name)}' is not supported");
    }

    private static bool IsMedicationStatementContainedInRetrievals(MedicationStatement medicationStatement, IEnumerable<MedicamentRetrieval> retrievals) =>
        retrievals.Select(retrieval => retrieval.Resource)
                  .OfType<MedicationStatement>()
                  .Any(ms => medicationStatement.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value
                          == ms.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value);

    private static bool IsMedicationDispenseContainedInRetrievals(MedicationDispense medicationDispense, IEnumerable<MedicamentRetrieval> retrievals) =>
        retrievals.Select(retrieval => retrieval.Resource)
                  .OfType<MedicationDispense>()
                  .Any(md => medicationDispense.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value
                          == md.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value);

    private static bool IsMedicationRequestContainedInRetrievals(MedicationRequest medicationRequest, IEnumerable<MedicamentRetrieval> retrievals) =>
        retrievals.Select(retrieval => retrieval.Resource)
                  .OfType<MedicationRequest>()
                  .Any(mr => medicationRequest.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value
                          == mr.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value);

    private static bool IsObservationContainedInRetrievals(Observation observation, IEnumerable<MedicamentRetrieval> retrievals) =>
        retrievals.Select(retrieval => retrieval.Resource)
                  .OfType<Observation>()
                  .Any(o => observation.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value
                         == o.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value);
``` 

I'd like to replace the 4 explicit methods by a single, generic one:
```c#
    private static bool IsIdentifialbleContainedInRetrievals<T>(T identifiable, IEnumerable<MedicamentRetrieval> retrievals)
        where T : IIdentifiable =>
        retrievals.Select(retrieval => retrieval.Resource)
                  .OfType<T>()
                  .Any(idf => identifiable.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value
                          == idf.Identifier.Single(id => id.System == FhirConstants.SystemIdentifier.ResourceIdentifier).Value);
```

where the new interface would be
```c#
public interface IIdentifiable {
  List<Hl7.Fhir.Model.Identifier> Identifier {get; set; }
}
```

Thanks for considering

**Comment by marcovisserFurore** [comment]
Hi @taconaut, thanks for suggestion. Always nice when we hear some ideas from the community. 

We will discuss this in our next sprint. It will not be so easy, because our resources are generated from the official FHIR standard and this will involve changes in the code generator. But we will come back to you after a few weeks.

**Comment by marcovisserFurore** [comment]
Maybe we can use the [FiveWs design pattern](https://www.hl7.org/fhir/fivews.html) to identity the identifier in a resource. We will investigate this in the current sprint (2022 4)

**Comment by taconaut** [comment]
@ewoutkramer To cover my use cases, the interface implementation (covering only the `Identifier` property) would be sufficient
```c#
public interface IIdentifiable {
  List<Hl7.Fhir.Model.Identifier> Identifier { get; set; }
}
```

**Comment by taconaut** [comment]
Hi,

I wanted to ask if you're still considering adding this? (as mentioned in the initial post I'm wondering if you couldn't use the same concept to provide an `IIdentifiable` interface as it is already done for the `IExtendable` interface?)

This would come in handy again for the UC I'm working on, where I have to retrieve the GLN of a practitioner and organization (if it has been set), which are referenced by the composition author (either as direct reference or through a `PractitionerRole`) for a [CH VACD Immunization Administration Composition](http://fhir.ch/ig/ch-vacd/StructureDefinition-ch-vacd-composition-immunization-administration.html).

It would allow me to remove (most) of the duplicated code
```c#
    private static string? GetVac23APractitionerGln(Bundle bundle)
    {
        var authorReferences = bundle.GetComposition().Author;

        if (TryGetPractitionerFromPractitionerReferences(bundle, authorReferences, out var practitionerGln))
        {
            return practitionerGln;
        }

        if (TryGetPractitionerFromPractitionerRoleReferences(bundle, authorReferences, out practitionerGln))
        {
            return practitionerGln;
        }

        return null;
    }

    private static bool TryGetPractitionerFromPractitionerReferences(Bundle bundle, IEnumerable<ResourceReference> practitionerReferences, out string? practitionerGln)
    {
        var practitioners = practitionerReferences.Select(practitionerReference => bundle.GetResourceByReference<Practitioner>(practitionerReference.Reference))
                                                  .OfType<Practitioner>()
                                                  .ToList();

        practitionerGln = GetPractitionerGln(practitioners);
        return practitionerGln != null;
    }

    private static bool TryGetPractitionerFromPractitionerRoleReferences(Bundle bundle, IEnumerable<ResourceReference> practitionerRoleReferences, out string? practitionerGln)
    {
        var practitioners = practitionerRoleReferences.Select(practitionerRoleReference => bundle.GetResourceByReference<PractitionerRole>(practitionerRoleReference.Reference))
                                                      .Where(practitionerRole => practitionerRole?.Practitioner != null)
                                                      .Select(practitionerRole => bundle.GetResourceByReference<Practitioner>(practitionerRole!.Practitioner!.Reference))
                                                      .OfType<Practitioner>()
                                                      .ToList();

        if (!practitioners.All(practitioner => practitioner.Identifier.Any(identifier => identifier.System == Constants.Systems.GlnIdentifierSystem)))
        {
            throw new MissingFhirPropertyException(typeof(Practitioner),
                                                   nameof(Practitioner.Identifier),
                                                   $"All practitioners referenced as author (Composition.Author -> PractitionerRole -> Practitioner) MUST contain an identifier with System={Constants.Systems.GlnIdentifierSystem}");
        }

        practitionerGln = GetPractitionerGln(practitioners);
        return practitionerGln != null;
    }

    private static string? GetPractitionerGln(IEnumerable<Practitioner> practitioners)
        => practitioners.SelectMany(practitioner => practitioner.Identifier)
                        .FirstOrDefault(identifier => identifier.System == Constants.Systems.GlnIdentifierSystem)
                       ?.Value;

    private static string? GetVac23AOrganizationGln(Bundle bundle)
    {
        var authorReferences = bundle.GetComposition().Author;

        if (TryGetOrganizationFromOrganizationReferences(bundle, authorReferences, out var organizationGln))
        {
            return organizationGln;
        }

        if (TryGetOrganizationFromPractitionerRoleReferences(bundle, authorReferences, out organizationGln))
        {
            return organizationGln;
        }

        return null;
    }

    private static bool TryGetOrganizationFromOrganizationReferences(Bundle bundle, IEnumerable<ResourceReference> organizationReferences, out string? organizationGln)
    {
        var organizations = organizationReferences.Select(organizationReference => bundle.GetResourceByReference<Organization>(organizationReference.Reference))
                                                  .OfType<Organization>()
                                                  .ToList();

        organizationGln = GetOrganizationGln(organizations);
        return organizationGln != null;
    }

    private static bool TryGetOrganizationFromPractitionerRoleReferences(Bundle bundle, IEnumerable<ResourceReference> practitionerRoleReferences, out string? organizationGln)
    {
        var organizations = practitionerRoleReferences.Select(practitionerRoleReference => bundle.GetResourceByReference<PractitionerRole>(practitionerRoleReference.Reference))
                                                      .Where(practitionerRole => practitionerRole?.Organization != null)
                                                      .Select(practitionerRole => bundle.GetResourceByReference<Organization>(practitionerRole!.Organization!.Reference))
                                                      .OfType<Organization>()
                                                      .ToList();

        organizationGln = GetOrganizationGln(organizations);
        return organizationGln != null;
    }

    private static string? GetOrganizationGln(IEnumerable<Organization> organizations)
        => organizations.SelectMany(organization => organization.Identifier)
                        .FirstOrDefault(identifier => identifier.System == Constants.Systems.GlnIdentifierSystem)
                       ?.Value;
```

**Comment by ewoutkramer** [comment]
We; have all intention to do it, but it keeps on being deprioritized. So, if you'd create a PR which adds IIdentifiable to the right resources (probably using partial classes, since you don't want to meddle with the generated stuff), we could integrate this quickly.

**Comment by taconaut** [comment]
@ewoutkramer you mean creating a partial class like https://github.com/taconaut/firely-net-sdk/pull/1 this for the 111 classes having this property?

**Comment by ewoutkramer** [comment]
There are 111 classes with an Identifier prop? Never realized we had so many. Nah, then we need to change the code generator.

**Comment by taconaut** [comment]
Yeah, most but not all of the models have these properties.  111 in R4 and 141 in R5. Having to do this manually would be tedious and lead to a maintenance nightmare.
Having this customizable in the code generator would be great to also support other interfaces like e.g.
```c#
public interface ICodeable {
  Hl7.Fhir.Model.CodeableConcept Code { get; set; }
}
```
which could be applied to 56 classes in R4.

**Comment by taconaut** [comment]
FYI I've started working on a postprocessor, which will have to be ran after the generation of the FHIR models, before the models are being committed to the GIT repo.

What it does:
1. Determine which interfaces have to be applied -> this is being determined using reflection by scanning one or more assemblies passed as command line parameter. Relevant classes are marked with a new attribute which can only be applied to interfaces.
2. Determine which classes an interface can be applied to -> this is being determined using reflection and comparing property and method signatures
3. Let classes determined in 2 extend the interface -> this is being done using Roslyn
-> The result is that the *.cs model files are modified in the file system to extend the interface(s) and can be committed to the GIT repo and ultimately can be built and released.

I'll make a PR once everything is ready (not sure when this will be as I got more urgent tasks).

Would this fit into the generation process?  You would have to call the new post-processor, which will live in `\build\tools` with parameters for 1) the assemblies containing the interface definitions 2) the assemblies containing the classes the interfaces will be applied to 3) the path to the source files.

**Comment by ewoutkramer** [comment]
If you'd build this straight into the generator, you could use the source data from which the code is generated, instead of doing the reflection after generation - we would not need post-processing!

Here's the repo https://github.com/microsoft/fhir-codegen,  look in the CsFirely2.cs class and find where we generated the classes and datatypes - it should be easy to detect a member called "Identifier" there.

**Comment by taconaut** [comment]
I've stayed with the post processor approach for different reasons. Hope it fits

**Comment by taconaut** [comment]
Hi @ewoutkramer, did you have the opportunity to have a look at the PR?
Running it, as it is meant to be ran during the generation process, is as simple as executing `.\build\tools\apply-interfaces-on-generated-classes.ps1`
Then GIT can be used to see the changes.

**Comment by taconaut** [comment]
Hope you had a good time :) No worries, just wanted to make sure it wasn't forgotten.
